### ISSUE TITLE
Sending and receiving multipart/form-data parts in JAX-RS

### ISSUE THREAD

**Issue by Charlotte-Holt** [OpenLiberty/docs#1273]
This is a new topic about how to configure a resource to receive multipart/form-data parts from an HTML form submission in JAX-RS. From SME: For the most part, users will know the concept of multi-part if they use it (or plan to use it) - they'll mainly just need to know how to implement it.

**Resource**: https://www.ibm.com/support/knowledgecenter/SSAW57_liberty/com.ibm.websphere.wlp.nd.multiplatform.doc/ae/twlp_jaxrs_multipart_formdata_from_html.html

**SME**: Andy McCright

**Comment by Charlotte-Holt** [comment]
Raised on the guides backlog: https://github.com/OpenLiberty/guides-common/issues/521

**Comment by andymc12** [comment]
Sorry to commandeer this issue, but now that we are adding the capability to send multipart payloads (see https://github.com/OpenLiberty/open-liberty/issues/16031 ), I'd like to use this issue to document both sending and receiving.

There is some information on sending multipart payloads in the beta blog post issue, https://github.com/OpenLiberty/open-liberty/issues/16249 - and I can help with drafting the docs and/or code samples if desired.  Thanks!

**Comment by andymc12** [comment]
Talking with @dmuelle  - I think we can resolve this issue with basically a lift-and-shift from the KC article ( https://www.ibm.com/support/knowledgecenter/SSEQTP_liberty/com.ibm.websphere.wlp.doc/ae/twlp_jaxrs_multipart_formdata_from_html.html ) with a few exceptions:

(1) The Java code example for receiving a multipart payload seems to be duplicated - we only need one example.
(2) That example should remove the `@Produces("multipart/form-data")` line since it is not producing a multipart response.
(3) We could probably "modernize" the code example - maybe something like this:

```
@POST
@Consumes("multipart/form-data")
public Response postFormData(List <IAttachment> attachments) throws IOException{
    InputStream stream = null;
    for (IAttachment attachment : attachments) {
         if (attachment == null) {
             continue;
         }
         stream = attachment.getDataHandler().getInputStream();
         String fileName = null;
         String contentType = attachment.getDataHandler().getContentType();
         if (contentType != null && contentType.startsWith("application/octet-stream")) {
            fileName = attachment.getDataHandler().getName();
         }
         if (fileName == null) {
             StringBuilder sb = new StringBuilder();
             BufferedReader br = new BufferedReader(new InputStreamReader(stream));
             String line = null;
             try {
                 while ((line = br.readLine()) != null) {
                     sb.append(line);
                 }
             } catch (IOException e) {
                 e.printStackTrace();
             } finally {
                 if (br != null) {
                     try {
                         br.close();
                     } catch (IOException e) {
                         e.printStackTrace();
                     }
                 }
             }
             System.out.println("Non-file attachment value: " + sb.toString());
         } else {
             //handle the file as you want
             File tempFile = new File(fileName);
             System.out.println("File: " + tempFile.getAbsolutePath());
         }
    }
    if (stream != null) {
        stream.close();
    }
    return Response.ok("test").build();
}
```

(4) We also need an example to cover the new ability to _send_ multipart payloads.  Maybe something like this:

```
List<IAttachment> attachments = new ArrayList<>();

attachments.add(AttachmentBuilder.newBuilder("fileid")
                                 .inputStream(new ByteArrayInputStream("person1234".getBytes()))
                                 .build());

attachments.add(AttachmentBuilder.newBuilder("description")
                                 .inputStream(new ByteArrayInputStream("XML file about person1234".getBytes()))
                                 .build());

File file = new File("/path/to/multipart-response/person.xml");
attachments.add(AttachmentBuilder.newBuilder("thefile")
                                 .inputStream("person.xml",new FileInputStream(file))
                                 .contentType(MediaType.APPLICATION_XML_TYPE)
                                 .build());

Client c = ClientBuilder.newClient();
WebTarget target = c.target("http://localhost:9080/myMultipartApp/person");
Response r = target.request(MediaType.TEXT_PLAIN)
                   .header("Content-Type", "multipart/form-data")
                   .post(Entity.entity(attachments, MediaType.MULTIPART_FORM_DATA_TYPE));

if(r.getStatus()==200 && "test".equals(r.readEntity(String.class)) {
    System.out.println("multipart upload succeeded");
}
```

**Comment by andymc12** [comment]
Oh - one other thing to mention (sorry - just remembered this as I was working on the serviceability aspect of this feature):
(5) The user should _not_ close the `InputStream` when sending multipart payloads - Liberty will do that for them after the data has been sent - and Liberty will throw an `IOException` if the stream is closed beforehand.  When _receiving_ multipart payloads, Liberty will _not_ close the `InputStream` because the server does not know when the user application is done processing it.

**Comment by dmuelle** [comment]
Hi @andymc12 - I've got a draft ready for your review:
[Send and receive multipart/form-data parts in JAX-RS (Draft)](https://draft-openlibertyio.mybluemix.net/docs/latest/send-receive-multipart-jaxrs.html)

The KC source was pretty sparse so I tried to add just a little more context. I imagine some of the names/relationships in the doc will need to be tuned up so please let me know if anything is incorrect, unclear, or awkwardly put.

Also, there was some info in the KC topic that was just kind of hanging out without much context. I moved it over but I need to understand the significance of it a little better:

> You can implement your JAX-RS resource method to receive the data in parts, so you can process these parts yourself, if needed. 

Does this relate directly to the "sending" code example? If so, how? If not, seems like we're vaguely telling the user something is possible but not giving any info about how or why they would want to do it. If this is information that is likely to be relevant to most scenarios, what else can we say about it?

> The originator of the HTML form POST submission can generate a Content-Transfer-Encoding header for one or more parts of the multipart payload. In these cases, if the header is of base64 or quoted-printable encoding type, the JAX-RS resource attempts to auto-decode the payload of each part according to the header.

I think I get what this is saying- does the code example support it? ie Using that method, can it process a Content-Transfer-Encoding header and is that a thing that is likely to be relevant?

**Comment by andymc12** [comment]
Hi @dmuelle - this looks good!  Here are a few comments:

* First sentence
> Multipart/form-data is a popular encoding type for HTTP requests that is used to send different types file data in a single request.

Could we change it to something like:
_Multipart/form-data is a popular content type for HTTP requests that can send different types of data in a single request._


* The second paragraph is text, but it is rendering in a code block - and all on one line, making it difficult to read.
* In that second paragraph, can you add inline code-formatting around "multipart/form-data" (i.e. `multipart/form-data`) and change _encoding_ type to _content_ (or media) type?
* At the end of the second paragraph, let's remove "GET" from the last sentence: "in response to an HTTP GET request." JAX-RS resource methods can return a multipart payload for any type of HTTP request.


* In the "Sending..." section, can you change this line:
> To send a multipart payload, you must send an array of objects as an instance of the List<IAttachment> method. Each object in that array represents a single attachment part. 

to something like:
_To send a multipart payload, you must send a list of `IAttachment` objects. Each object in that list represents a single attachment part._


* And can you change this line:
> In this example, the JAX-RS resource implements a method to send an XML file, along with an ID and description for the file, to a client in the same POST request.

to something like:
_In this example, the client sends a POST request to a remote service with three parts: an XML file, along with an ID and description for the file._


* In the following sentence, would you remove "class" - or change it to "object" or "instance" :
> If you are sending multipart payloads, do not close the InputStream class.




* In the link at the bottom, it points to "#page-4" of the RFC and probably should just point to the top of the page - so instead of `https://tools.ietf.org/html/rfc7578#page-4`, it should be `https://tools.ietf.org/html/rfc7578`.



General thought: The HTML form code does the same thing as the "sending" Java code.  The HTML approach would be for developing a web page that allows human users to send multipart data (though as far as they know they are just filling out a form - and possibly uploading one or more files).  The Java code approach would be for automating the same request - for example, it might be part of a micro-service that automatically submits a purchase order when it receives a certain "low inventory" event.
Since the two things do basically the same thing, maybe we could move the "Send" section to under the "Receiving" section - and introduce it by saying something like "now, if we wanted to send that same data from a Java client, we would do something like ...".  What do you think?


Related to:
> You can implement your JAX-RS resource method to receive the data in parts, so you can process these parts yourself, if needed.

The term "parts" is synonymous with "attachment".  "Attachment" is the older term, but "part" just means that it is one part of the multipart payload.  I think this sentence is mostly filler... kinda like "since a multipart payload is composed of multiple parts, you can process them in a list of parts (aka attachments)".  Maybe we could rephrase it to something like:
_Liberty will pass a `List<IAttachment>` to your resource method allowing you to process each part of the multipart payload._


Regarding:
> The originator of the HTML form POST submission can generate a Content-Transfer-Encoding header for one or more parts of the multipart payload. In these cases, if the header is of base64 or quoted-printable encoding type, the JAX-RS resource attempts to auto-decode the payload of each part according to the header.

Your concerns are valid... I'm not an expert on what HTML multipart forms can do, but I don't think it can set headers on individual parts - and if it can, I don't think the HTML sample code shows how to do it...  We should communicate that headers on individual parts are possible - and that JAX-RS attempts to auto-decode the payload of each part.  Maybe we could rephrase it like:
_It is possible for clients to send headers specific to each part of the multipart payload. If a part contains the `Content-Transfer-Encoding` header that is set to base64 or a quoted-printable encoding type, the runtime will attempt to auto-decode the payload of that part._

**Comment by dmuelle** [comment]
Hi @andymc12 

I've made the requested updates per your comment:

https://draft-openlibertyio.mybluemix.net/docs/latest/send-receive-multipart-jaxrs.html

I kept the HTML form example for now, basically to help illustrate the use-case for JAX-RS with multpart content (automation). See if it makes sense. If it seems unnecessary or out-of-place, I can remove it.

**Comment by andymc12** [comment]
HI @dmuelle - the updates look perfect!  I like how you organized the "sending" section with the HTML first and the Java code second.  I think it looks more concise than the original KC doc (which has less content!).  Well done!  Thanks!

**Comment by ManasiGandhi** [comment]
# Content Review

@dmuelle the topic looks good. A few suggestions

## Send and receive multipart/form-data parts in JAX-RS

- [x] 	“JAX-RS resources can implement the `multipart/form-data` content type to receive form data such as this from a client.” Could you reword the sentence to avoid the word “this”?

## Send multipart/form-data parts from JAX-RS resources

- [x] “Each object in that list represents a single attachment part. In the following example, a JAX-RS client sends a `POST` request to a remote service. The request has three parts: an XML file, an ID string for the file, and a text description of the file.” In these sentences the parts are differently mentioned as “attachment part” and “three parts”. Could you replace “attachment part” with “part” for consistency?

- [x] Also, could you elaborate more and on `IAttachment` object and `InputStream` object to make the concept clearer? `InputStream` object: Could you elaborate on why the user would want to close the InputStream object, before asking them not to do so?

- [x] Instead of `IOException` exception, is it possible to write IOException error. Here it is mentioned as an error in this support doc https://www.ibm.com/support/pages/javaioioexception-error-when-opening-or-adding-attachments-clearquest-web. Though, use your own discretion.

## Receive multipart/form-data parts with JAX-RS resources

- [x] Delete the comment in the code example “//handle the file as you want”.

- [ ] Is there an alternative way to document  `List<IAttachment>` instance without the angle brackets?

- [x] “However, the `Content-Transfer-Encoding` header must be set to either the `base64` or a `quoted-printable` encoding type.” Could you elaborate more on this?

- [x] In general, `multipart/form-data` is referred to as content type, part, and payload. Is it possible to refer to it consistently as part?

**Comment by andymc12** [comment]
Technical answers (please feel free to re-word): 

> Also, could you elaborate more and on IAttachment object and InputStream object to make the concept clearer? 

`IAttachment` represents one part of the multipart payload.  "Attachment" is an older term, but is consistent with our existing APIs.  Somehow the doc should indicate that "IAttachment" is the Java code equivalent of "part".

> InputStream object: Could you elaborate on why the user would want to close the InputStream object, before asking them not to do so?

Closing objects when you are done with them is a common coding pattern - for anything that could leave resources open, not just streams.  Users who follow good coding practices will naturally want to close any object that they "open" to avoid things like memory or file handle leaks.  It's important in this case that they do not close the stream - otherwise it may cause errors with operation.

> Is there an alternative way to document `List<IAttachment>` instance without the angle brackets?

Not really... the complete multipart payload is represented (in Java code) as `List<IAttachment>`. In text, I suppose you could describe it as a list of attachments.

> In general, `multipart/form-data` is referred to as content type, part, and payload. Is it possible to refer to it consistently as part?

`multipart/form-data` is the content type of the HTTP request. When an HTTP request is of type, `multipart/form-data`, it will contain a payload that includes one or more parts.  (Other HTTP requests might not contain a payload at all, and still others may contain a payload but only with a single entity. Multipart requests are unique in that they contain multiple parts.)

HTH

**Comment by dmuelle** [comment]
Thanks @andymc12 - very helpful

@ManasiGandhi thanks for reviewing, I made the following changes per yur review and based on Andy's feedback
fix style issues mentioned above:
- removed "this"
- made references to "multipart payloads" vs "`multipart/form-data` parts" consistent
- per Andy's comment- I added a sentence for context around `IAttachment` but did not go into detail re closing InputStream as "Closing objects when you are done with them is a common coding pattern", likely to be familiar to developers already
- exception-->error
- deleted specified code comment
- left `List<IAttachment>` as is- though we generally do not include brackets around XML elements in text references, in this case they are part of a syntax that is specific to this instance
- rephrased reference to Content-transfer-encoding header and added a link to the spec

**Comment by andymc12** [comment]
@dmuelle @ManasiGandhi - the latest changes look to me.  Thanks!

**Comment by Rwalls1** [comment]
# Peer review

## Introduction

I think the phrase “in the JAX-RS resource code” should be changed to reference a more specific location such as “in the JAX-RS API” to avoid possible confusion as I don’t think it’s clear what “resource code” is referring to.

## Send multipart/form-data parts from JAX-RS resources
- I think you should change "The following example shows an HTML form that sends an HTTP request in three parts, each of which specifies a particular content type:” to something like, “The following example shows an HTML form that sends an HTTP request in three parts that each specify a content type” to make the sentence simpler and avoid the use of unnecessary additional clause. I also think that you should specify which part of the request form specifies the “particular content type”. So, I think maybe add a sentence that says something like, “The `type` attribute in the `input` element specifies the content type for the HTTP request”, just to maintain clarity
- I think you should also add a bit more detail for the config example of the `POST` request. The included example is very long and I think it might be difficult to understand without a bit more context. I think you should briefly describe where the three parts of the request are in the example.
- Without additional context, the phrase, “do not close the `InputStream` object” doesn’t add much value because it is not clearly explained how the `InputStream` object works or what it does. I think you should either provide more detail or provide a reference to resource that might provide more explanation. I think you should apply a similar approach when describing the config example in the “Receive multipart/form-data parts with JAX-RS resources” section, which is also very lengthy 
## Receive multipart/form-data parts with JAX-RS resources
I think the phrase “in your application code” should be changed to something like “in your API” to highlight a more specific location and avoid possible confusion.

**Comment by dmuelle** [comment]
Thanks for reviewing @Rwalls1 - I made the following changes per your review:

- the JAX-RS resource code-->the JAX-RS resource methods (per Andy)
- revised intro to HTTP example to be more explicit
- added more detail to POST request config example, InputStream reference and Receiving explanation
- "I think the phrase “in your application code” should be changed to something like “in your API” to highlight a more specific location and avoid possible confusion." Per Andy's suggestion I left this as is. The example doesn't show API code, but application code that makes use of the JAX-RS API. This code would be in the application that runs on OL and wants to receive multipart payload

**Comment by dmuelle** [comment]
comments from @andymc12 on slack:

Personally, I like the text as it is better - or maybe just tweak it slightly to:
"You can send and receive multipart/form-data parts with JAX-RS resources in Open Liberty by enabling the Jakarta RESTful Services feature and implementing the necessary code in the JAX-RS resource methods."
(just changing Java RESTful to Jakarta RESTful and swapping code and methods)
For:
"You must explicitly close the object in your application code."
I wouldn't change that at all. We are telling the user not to close the object - and the only place that they could close it (and that they have control over) would be their application code.

For the "receive" example:

I think you've got the main thing covered (that the users need their resource method to have a List<IAttachment> parameter) - everything else is just how the user wants to process each attachment/part.  So maybe something like this:
With this configuration, Open Liberty can pass a List<IAttachment> instance to your resource method so that you can process each part of the multipart payload.  How you process the list of IAttachment parts is up to you, but in the example above, the postFormData method iterates over each part and checks to see whether that part contains a file or not. It then prints the contents to the console output (System.out).  Notice also, that you can check the part's content type to determine how best to handle the part's content.
I'm thinking that we might want to change the code a little... right now it determines whether the IAttachment contains a file or not based on the content type.  That might be perfectly valid for scenarios where the server is aware of it's clients and knows that they will only send files with the application/octet-stream content type - but in other scenarios, it is perfectly valid for a client to send a file with a different content type.  The main thing for users to understand with that is that they can check to see what the content type is.  Maybe we could change this code:
         String fileName = null;
         String contentType = attachment.getDataHandler().getContentType();
         if (contentType != null && contentType.startsWith("application/octet-stream")) {
            fileName = attachment.getDataHandler().getName();
         }
to:
         String contentType = attachment.getDataHandler().getContentType();
         System.out.println("Content-type: " + contentType);
         String fileName = attachment.getDataHandler().getName();

**Comment by dmuelle** [comment]
added to KC spreadsheet and posted a link from the KC topic to the updated OL page. Closing this isssue
